### ISSUE TITLE
`sync` in shell wrapper really glugs the server when lots of wf submitted

### ISSUE THREAD

**Issue by delocalizer** [broadinstitute/cromwell#2057]
The [sync here](https://github.com/broadinstitute/cromwell/blob/fac784dd4078b8cc12fb4ca6c9abdbb05072990b/backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala#L192) is proving very costly for our SFS cromwell server...

`sync` flushes _everything_ to disk. Depending on how much memory you've got, what else is happening on the box, and type of filesystem, that can be a really expensive call. For example on one of our servers  it typically takes about 10 seconds to return and uses 100% of 1 core in system cpu. When you run cromwell on that box and throw a thousand workflows at it, you end up with load like this:
```
top - 16:16:42 up 196 days,  6:34, 17 users,  load average: 806.48, 618.79, 413.58
Tasks: 4596 total, 829 running, 3766 sleeping,   1 stopped,   0 zombie
%Cpu(s):  1.4 us,  0.5 sy,  0.0 ni, 98.1 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem : 26386596+total, 15019235+free, 25326680 used, 88346944 buff/cache
KiB Swap: 67108860 total, 63149068 free,  3959792 used. 21917659+avail Mem 

   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                       
 34965 cromwel+  20   0 42.622g 896064  20048 S   9.5  0.3   9:33.71 java                                          
 88343 cromwell  20   0  107892    360    284 R   7.3  0.0   0:28.21 sync                                          
 88838 cromwell  20   0  107892    360    284 R   7.3  0.0   0:27.78 sync                                          
 89911 cromwell  20   0  107892    360    284 R   7.3  0.0   0:23.89 sync                                          
 82582 cromwell  20   0  107892    356    284 R   6.7  0.0   0:31.86 sync                                          
 84039 cromwell  20   0  107892    356    284 R   6.7  0.0   0:32.20 sync                                          
 84917 cromwell  20   0  107892    360    284 R   6.7  0.0   0:28.46 sync                                          
 84962 cromwell  20   0  107892    356    284 R   6.7  0.0   0:28.67 sync                                          
 85475 cromwell  20   0  107892    356    284 R   6.7  0.0   0:28.88 sync                                          
 88464 cromwell  20   0  107892    356    284 R   6.7  0.0   0:28.14 sync                                          
 88465 cromwell  20   0  107892    360    284 R   6.7  0.0   0:27.49 sync                                          
 89112 cromwell  20   0  107892    356    284 R   6.7  0.0   0:27.73 sync                                          
 89398 cromwell  20   0  107892    360    284 R   6.7  0.0   0:22.13 sync                                          
 91615 cromwell  20   0  107892    356    284 R   6.7  0.0   0:12.59 sync                                          
 85206 cromwell  20   0  107892    356    284 R   6.1  0.0   0:27.26 sync                                          
 85628 cromwell  20   0  107892    356    284 R   6.1  0.0   0:28.59 sync                                          
 86528 cromwell  20   0  107892    360    284 R   6.1  0.0   0:27.53 sync                                          
 86833 cromwell  20   0  107892    360    284 R   6.1  0.0   0:27.62 sync                                          
 86848 cromwell  20   0  107892    356    284 R   6.1  0.0   0:28.30 sync                                          
 87034 cromwell  20   0  107892    356    284 R   6.1  0.0   0:27.50 sync                                          
 87256 cromwell  20   0  107892    360    284 R   6.1  0.0   0:27.27 sync                                          
 87465 cromwell  20   0  107892    360    284 R   6.1  0.0   0:28.04 sync                                          
 89692 cromwell  20   0  107892    360    284 R   6.1  0.0   0:22.21 sync                                          
 89976 cromwell  20   0  107892    356    284 R   6.1  0.0   0:23.04 sync                                          
 90467 cromwell  20   0  107892    356    284 R   6.1  0.0   0:19.82 sync                                          
 90988 cromwell  20   0  107892    360    284 R   6.1  0.0   0:13.32 sync                                          
 91427 cromwell  20   0  107892    356    284 R   6.1  0.0   0:13.01 sync                                          
 91645 cromwell  20   0  107892    356    284 R   6.1  0.0   0:12.66 sync                                          
 91971 cromwell  20   0  107892    360    284 R   6.1  0.0   0:11.03 sync                                          
 91973 cromwell  20   0  107892    360    284 R   6.1  0.0   0:10.64 sync                                          
 92276 cromwell  20   0  107892    356    284 R   6.1  0.0   0:09.48 sync                                          
 92612 cromwell  20   0  107892    356    284 R   6.1  0.0   0:08.25 sync                                          
 92756 cromwell  20   0  107892    356    284 R   6.1  0.0   0:08.03 sync                                          
 92834 cromwell  20   0  107892    356    284 R   6.1  0.0   0:07.79 sync                                          
 92959 cromwell  20   0  107892    356    284 R   6.1  0.0   0:07.61 sync 
...
```
which is not great

I don't have a good alternative, but for mine, `sync` isn't the right solution to issue mentioned in [this PR](https://github.com/broadinstitute/cromwell/pull/1974) - I don't even think it absolutely guarantees disk write, for example if you have hardware disk caching enabled.

For the time being I'm going to try a build without the sync, and get some numbers to show load avg comparison with and without

**Comment by delocalizer** [comment]
![load_avg_with_and_without_sync](https://cloud.githubusercontent.com/assets/817809/23653377/0b318f3e-0378-11e7-9b24-a2cbe2b3e2dd.png)

This is a plot of load average on our machine that hosts cromwell running as server - the box has 48 cores and 256G RAM. Mounted filesystems are local disk, an NFS share, and Lustre filesystem that hosts cromwell executions directory. The chart covers two identical submissions of a batch of about 1500 workflows. The period on the left from 5:45 to 7:30 with load peaking up around 1700 is running with release 25 jar. The period on the right from 9:15 to 10:30 is running with a build that is identical to release 25 but without [this sync](https://github.com/broadinstitute/cromwell/blob/fac784dd4078b8cc12fb4ca6c9abdbb05072990b/backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala#L192).
It's a big difference.

**Comment by kcibul** [comment]
That plot is the load on the Cromwell server itself?  

From what's you've said, and the top output, it looks like your tasks are exeucting on that server as well.  You mentioned that you're running SFS, can you tell us more about your configuration?  Are you dispatching to PBS but the execution host is the cromwell server itself (where the sync commands are running)?

**Comment by delocalizer** [comment]
Good point - that is the load on cromwell server itself; default backend is PBS (very like SGE) but some of the simple tasks in the workflow (`mkdir`, `uuidgen`) specify `backend: "Local"` in the WDL `runtime` block.

**Comment by delocalizer** [comment]
Regardless of where the task is executing though, that `sync` is expensive, and the more cores and memory you have on execution nodes, and the less work the `command` is actually doing, the worse things become when lots of tasks get allocated to a given node.
Simple "echo Hello, World" tasks occupy one of our 56-core execution nodes for about 30 seconds when 56 of them are allocated to a node, nearly the whole time spent in sync. 
The answer could well be "then stop doing that", but I hope not!

**Comment by geoffjentry** [comment]
@delocalizer @kcibul we talked about this internally. As background we went down this path as our integration tests were frequently failing in travis - hte output files would be empty or incomplete. 

It was noted that our tests use a lot of `echo` and `cat` and are quite short, so the theory is we're running into [this](https://www.turnkeylinux.org/blog/unix-buffering).

*if* that turns out to be the culprit (and it does make a lot of sense) one could either take the stance that tools need to be well formed and have properly flushed, or we could try to bake something into our controller bash script (which IMO adding so much stuff to that bash script is a bomb waiting to happen, but ....), some [ideas](http://serverfault.com/questions/294218/is-there-a-way-to-redirect-output-to-a-file-without-buffering-on-unix-linux) are in that link.

@kcibul what's your reaction to the above? does it ring true or still seem fishy?

**Comment by kcibul** [comment]
I'll defer to @davidbernick as to the fishiness

**Comment by mcovarr** [comment]
Might we experiment with [unbuffer](http://unix.stackexchange.com/questions/25372/turn-off-buffering-in-pipe) to test this hypothesis?

**Comment by geoffjentry** [comment]
@mcovarr I did mention doing such a thing in slack ;) The big problem is it's hard to prove the absence of the problem, we can only demonstrate that it becomes better.

**Comment by Horneth** [comment]
I like to think that we're constantly proving the absence of all the problems we don't have :p

**Comment by delocalizer** [comment]
An update...
It looks like the performance of `sync` — run on command line entirely outside the context of cromwell —  that we see on our box that happens to be running cromwell is atypical. Other machines in our compute cluster with same OS and similar amounts of memory execute `sync` in milliseconds, the cromwell server box takes up to a second. we're still trying to identify the cause; there are some configuration differences in the machines, specifically the number, type and state of the mounted filesystems, as well as differences in the applications running in background, but as it stands I'm happy to call this as not a cromwell performance issue, so much as a potential performance issue running cromwell in certain environments.

**Comment by geoffjentry** [comment]
@delocalizer We're starting to consider that the issue is in tooling, specifically in the tools we use for our integration tests. Since you are as far as I know the largest user of the shared file system backend(s), to what degree do you trust that tools are flushing when they're complete?

**Comment by delocalizer** [comment]
That's tricky to answer... we use a huge number of tools and the behaviour (good or bad) of only a small number of them is well characterized, or known to us anyway. I'd say by default we don't trust anything to flush its output buffers until it's proven that it does. And then there's the issue of process pipe and redirection buffering; as soon as WDL-writers put `|` or `>` in commands then you've introduced an OS-level io buffer and previously well-behaved tools that were write in unbuffered or line-buffered mode to stdout now have 4k buffers. So a lot of odd things certainly can and will happen in the `command { }` block.

**Comment by delocalizer** [comment]
Fixed by #2160
